File: Bug FFI/kuwazy_engine/wrapper/javascript/math/Vector3.mjs
import * as FFI from "../FFI.mjs";

const registry = new FinalizationRegistry((pointer) => {
    FFI.vector3_free(pointer);
});

export default class Vector3{
    #ptr;
    constructor (x = 0, y = 0, z = 0){
        this.#ptr = FFI.vector3_new(x, y, z);
        registry.register(this, this.#ptr);
    }

    get_pointer(){
        return this.#ptr;
    }
    
    get x(){
        return FFI.vector3_get_x(this.#ptr);
    }
    set x(new_x){
        FFI.vector3_set_x(this.#ptr, new_x);
    }

    get y(){
        return FFI.vector3_get_y(this.#ptr);
    }
    set y(new_y){
        FFI.vector3_set_y(this.#ptr, new_y);
    }

    get z(){
        return FFI.vector3_get_z(this.#ptr);
    }
    set z(new_z){
        FFI.vector3_set_z(this.#ptr, new_z);
    }

    clone(){
        return new Vector3(this.x, this.y, this.z);
    }

    copy(vector3_source){
        FFI.vector3_copy(this.#ptr, vector3_source.get_pointer());
        return this;
    }

    set(x, y, z){
        FFI.vector3_set(this.#ptr, x, y, z);
        return this;
    }

    add(vector3){
        FFI.vector3_add(this.#ptr, vector3.get_pointer());
        return this;
    }

    subtract(vector3){
        FFI.vector3_subtract(this.#ptr, vector3.get_pointer());
        return this;
    }

    multiply(vector3){
        FFI.vector3_multiply(this.#ptr, vector3.get_pointer());
        return this;
    }

    divide(vector3){
        FFI.vector3_divide(this.#ptr, vector3.get_pointer());
        return this;
    }

    modulus(vector3){
        FFI.vector3_modulus(this.#ptr, vector3.get_pointer());
        return this;
    }

    pow(vector3){
        this.x **= vector3.x;
        this.y **= vector3.y;
        this.z **= vector3.z;
        return this;
    }

    add_scalar(scalar){
        FFI.vector3_add_scalar(this.#ptr, scalar);
        return this;
    }

    multiply_scalar(scalar){
        FFI.vector3_multiply_scalar(this.#ptr, scalar);
        return this;
    }

    divide_scalar(scalar){
        FFI.vector3_divide_scalar(this.#ptr, scalar);
        return this;
    }

    inverse(){
        FFI.vector3_inverse(this.#ptr);
        return this;
    }

    cross(vector3){
        FFI.vector3_cross(this.#ptr, vector3.get_pointer());
        return this;
    }

    dot(vector3){
        return FFI.vector3_dot(this.#ptr, vector3.get_pointer());
    }

    is_equal_to(vector3){
        return FFI.vector3_is_equal(this.#ptr, vector3.get_pointer());
    }

    linear_interpolate(to, weight){
        FFI.vector3_linear_interpolate(this.#ptr, to.get_pointer(), weight);
        return this;
    }

    distance_squared_to(vector3){
        return FFI.vector3_distance_squared(this.#ptr, vector3.get_pointer());
    }

    distance_to(vector3){
        return FFI.vector3_distance(this.#ptr, vector3.get_pointer());
    }

    direction_to(vector3){
        FFI.vector3_direction(this.#ptr, vector3.get_pointer());
        return this;
    }

    length(){
        return FFI.vector3_length(this.#ptr);
    }

    length_squared(){
        return FFI.vector3_length_squared(this.#ptr);
    }

    normalize(){
        FFI.vector3_normalize(this.#ptr);
        return this;
    }

    angle_to(vector3){
        return FFI.vector3_angle(this.#ptr, vector3.get_pointer());
    }
}

File: Bug FFI/kuwazy_engine/src/math/vector3.c
#include <stdbool.h>
#include <math.h>
#include <stdlib.h>
#include "vector3.h"
#include "utils.h"

Vector3 *vector3_new(const float x, const float y, const float z){
    Vector3 *vec3 = (Vector3 *)malloc(sizeof(Vector3));
    if (vec3){
        vector3_set(vec3, x, y, z);
    }
    return vec3;
}

void vector3_free(Vector3 *vector3){
    free(vector3);
}

Vector3 *vector3_clone(const Vector3 *vector3){
    return vector3_new(vector3->x, vector3->y, vector3->z);
}

void vector3_copy(Vector3 *a, const Vector3 *b){
    vector3_set(a, b->x, b->y, b->z);
}

void vector3_set(Vector3 *vector3, const float x, const float y, const float z){
    vector3->x = x;
    vector3->y = y;
    vector3->z = z;
}

float vector3_get_x(const Vector3 *vector3){
    return vector3->x;
}
void vector3_set_x(Vector3 *vector3, const float value){
    vector3->x = value;
}

float vector3_get_y(const Vector3 *vector3){
    return vector3->y;
}
void vector3_set_y(Vector3 *vector3, const float value){
    vector3->y = value;
}

float vector3_get_z(const Vector3 *vector3){
    return vector3->z;
}
void vector3_set_z(Vector3 *vector3, const float value){
    vector3->z = value;
}

void vector3_add(Vector3 *a, const Vector3 *b){
    a->x += b->x;
    a->y += b->y;
    a->z += b->z;
}

void vector3_subtract(Vector3 *a, const Vector3 *b){
    a->x -= b->x;
    a->y -= b->y;
    a->z -= b->z;
}

void vector3_multiply(Vector3 *a, const Vector3 *b){
    a->x *= b->x;
    a->y *= b->y;
    a->z *= b->z;
}

void vector3_divide(Vector3 *a, const Vector3 *b){
    a->x /= b->x;
    a->y /= b->y;
    a->z /= b->z;
}

void vector3_modulus(Vector3 *a, const Vector3 *b){
    a->x = fmod(a->x, b->x);
    a->y = fmod(a->y, b->y);
    a->z = fmod(a->z, b->z);
}

void vector3_add_scalar(Vector3 *vector3, const float scalar){
    vector3->x += scalar;
    vector3->y += scalar;
    vector3->z += scalar;
}

void vector3_multiply_scalar(Vector3 *vector3, const float scalar){
    vector3->x *= scalar;
    vector3->y *= scalar;
    vector3->z *= scalar;
}

void vector3_divide_scalar(Vector3 *vector3, const float scalar){
    if (scalar){
        vector3->x /= scalar;
        vector3->y /= scalar;
        vector3->z /= scalar;
    }
}

void vector3_inverse(Vector3 *vector3){
    if (vector3->x && vector3->y && vector3->z){
        vector3->x = 1 / vector3->x;
        vector3->y = 1 / vector3->y;
        vector3->z = 1 / vector3->z;
    }
}

void vector3_cross(Vector3 *a, const Vector3 *b){
    const float x = a->y * b->z - a->z * b->y;
    const float y = a->z * b->x - a->x * b->z;
    const float z = a->x * b->y - a->y * b->z;
    vector3_set(a, x, y, z);
}

float vector3_dot(const Vector3 *a, const Vector3 *b){
    return a->x * b->x + a->y * b->y + a->z * b->z;
}

bool vector3_is_equal(const Vector3 *a, const Vector3 *b){
    return a->x == b->x && a->y == b->y && a->z == b->z;
}

void vector3_linear_interpolate(Vector3 *a, const Vector3 *b, const float weight){
    a->x = lerp(a->x, b->x, weight);
    a->y = lerp(a->y, b->y, weight);
    a->z = lerp(a->z, b->z, weight);
}

float vector3_distance_squared(const Vector3 *a, const Vector3 *b){
    const float x = a->x - b->x;
    const float y = a->y - b->y;
    const float z = a->z - b->z;

    return x * x + y * y + z * z;
}

float vector3_distance(const Vector3 *a, const Vector3 *b){
    return sqrtf(vector3_distance_squared(a, b));
}

float vector3_length_squared(const Vector3 *a){
    return vector3_dot(a, a);
}

float vector3_length(const Vector3 *a){
    return sqrtf(vector3_length_squared(a));
}

void vector3_direction(Vector3 *a, const Vector3 *b){
    vector3_set(
        a,
        b->x - a->x,
        b->y - a->y,
        b->z - a->z
    );
    vector3_normalize(a);
}

void vector3_normalize(Vector3 *vector3){
    if (!(vector3->x || vector3->y || vector3->z)){
        vector3_set(vector3, 0, 0, 0);
    }

    const float length = vector3_length(vector3);
    vector3_divide_scalar(vector3, length);
}

float vector3_angle(const Vector3 *a, const Vector3 *b){
    Vector3 *a_copied = vector3_clone(a);
    vector3_cross(a_copied, b);

    const float result = atan2f(vector3_length(a_copied), vector3_dot(a, b));

    vector3_free(a_copied);

    return result;
}